pkg/server/aggregatingcrdversiondiscovery/server.go

```diff
@@ -0,0 +1,293 @@
+	"strings"
+
+	autoscaling "k8s.io/api/autoscaling/v1"
+	apiextensionshelpers "k8s.io/apiextensions-apiserver/pkg/apihelpers"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apiserver/pkg/endpoints/discovery"
+	genericapirequest "k8s.io/apiserver/pkg/endpoints/request"
+	genericapiserver "k8s.io/apiserver/pkg/server"
+	"k8s.io/apiserver/pkg/warning"
+
+	"github.com/kcp-dev/logicalcluster/v3"
+
+	"github.com/kcp-dev/kcp/pkg/indexers"
+	apisv1alpha2 "github.com/kcp-dev/kcp/sdk/apis/apis/v1alpha2"
+)
+
+var (
+	scheme = runtime.NewScheme()
+	codecs = serializer.NewCodecFactory(scheme)
+
+	unversionedVersion = schema.GroupVersion{Group: "", Version: "v1"}
+	unversionedTypes   = []runtime.Object{
+		&metav1.APIResourceList{},
+	}
+)
+
+func init() {
+	// we need to add the options to empty v1
+	metav1.AddToGroupVersion(scheme, schema.GroupVersion{Group: "", Version: "v1"})
+
+	scheme.AddUnversionedTypes(unversionedVersion, unversionedTypes...)
+}
+
+type Server struct {
+	GenericAPIServer *genericapiserver.GenericAPIServer
+	Extra            *ExtraConfig
+	delegate         http.Handler
+
+	verbsProvider *storageAwareResourceVerbsProvider
+
+	getCRD             func(cluster logicalcluster.Name, name string) (*apiextensionsv1.CustomResourceDefinition, error)
+	getAPIExportByPath func(clusterPath logicalcluster.Path, name string) (*apisv1alpha2.APIExport, error)
+}
+
+func NewServer(c CompletedConfig, delegationTarget genericapiserver.DelegationTarget) (*Server, error) {
+	s := &Server{
+		Extra:    c.Extra,
+		delegate: delegationTarget.UnprotectedHandler(),
+
+		getCRD: func(clusterName logicalcluster.Name, name string) (*apiextensionsv1.CustomResourceDefinition, error) {
+			return c.Extra.CRDLister.Lister().Cluster(clusterName).Get(name)
+		},
+		getAPIExportByPath: func(clusterPath logicalcluster.Path, name string) (*apisv1alpha2.APIExport, error) {
+			return indexers.ByPathAndNameWithFallback[*apisv1alpha2.APIExport](
+				apisv1alpha2.Resource("apiexports"),
+				c.Extra.LocalAPIExportInformer.Informer().GetIndexer(),
+				c.Extra.GlobalAPIExportInformer.Informer().GetIndexer(),
+				clusterPath,
+				name,
+			)
+		},
+	}
+
+	s.verbsProvider = &storageAwareResourceVerbsProvider{
+		getAPIExportByPath: s.getAPIExportByPath,
+		getAPIExportsByVirtualResourceIdentity: func(vrIdentity string) ([]*apisv1alpha2.APIExport, error) {
+			return indexers.ByIndexWithFallback[*apisv1alpha2.APIExport](
+				c.Extra.LocalAPIExportInformer.Informer().GetIndexer(),
+				c.Extra.GlobalAPIExportInformer.Informer().GetIndexer(),
+				indexers.APIExportByVirtualResourceIdentities,
+				vrIdentity,
+			)
+		},
+
+		knownVirtualResourceVerbs: map[string][]string{
```

Comment:
Feel like this need code comment why we doing this verbProvider thingy. I know we talked about this but other readers might not be aware of reasons.